### DISCUSSION TITLE
Accesing root value in nested validator

### DISCUSSION THREAD

**Discussion by richin13** [pydantic/pydantic#7672] [Question]
Hello Pydantic community!

I have a problem writing a pydantic validator that requires accessing the root value in a nested validator. The example I have so far looks like this:

```py
"""This is a minimal example of a pydantic model with a nested model."""
import pydantic


class NestedData(pydantic.BaseModel):
    """Nested data model."""

    a: int
    b: int

    @pydantic.field_validator("a")
    @classmethod
    def validate_a(cls, a: int, info: pydantic.ValidationInfo) -> int:
        """If root is 'first', a must be negative."""
        #: ??? How to get root model name? (first, second, _any_)
        if cls.ROOT_MODEL_VALUE == "first" a > 0:
            raise ValueError("first.a must be negative")
        return a


Data = pydantic.RootModel[dict[str, NestedData]]


def main():
    """Main entrypoint."""
    #: This should fail: first.a is positive
    Data.model_validate(
        {"first": {"a": 1, "b": 2}, "second": {"a": 1, "b": 2}}
    )
    #: This should pass. We don't cara about second.a
    Data.model_validate(
        {"first": {"a": -1, "b": 2}, "second": {"a": 1, "b": 2}}
    )


if __name__ == "__main__":
    main()

```

Is there an equivalent to `ROOT_MODEL_VALUE` in my example that would help me write my validator? I'm constrained by the python version (3.8) in case the most obvious solution would be using `Annotated` on the `Data = pydantic.RootModel[dict[str, NestedData]]` 🤔

**Comment by Ezequie1Sc** [comment]
I don't think there is a ROOT_MODEL_VALUE equivalent available inside NestedData. The nested model is validated independently, so the "first" / "second" dictionary key is not part of ValidationInfo.data.

Since the validation depends on the key of the outer dictionary, I would move this check to the RootModel:

import pydantic


class NestedData(pydantic.BaseModel):
    a: int
    b: int


class Data(pydantic.RootModel[dict[str, NestedData]]):

    @pydantic.model_validator(mode="after")
    def validate_data(self):
        first = self.root.get("first")

        if first is not None and first.a > 0:
            raise ValueError("first.a must be negative")

        return self


# This raises a validation error
Data.model_validate({
    "first": {"a": 1, "b": 2},
    "second": {"a": 1, "b": 2},
})

# This passes
Data.model_validate({
    "first": {"a": -1, "b": 2},
    "second": {"a": 1, "b": 2},
})

This keeps NestedData independent of the name of the key in the parent model.

If the validation really needs to stay inside NestedData, another option is to pass the parent key through Pydantic's validation context and access it with info.context inside the nested validator
